cfme/utils/ssh.py

```diff
@@ -842,3 +851,42 @@ def keygen():
     with _ssh_pubkey_file.open('w') as f:
         f.write("{} {} {}\n".format(pub.get_name(), pub.get_base64(),
             'autogenerated cfme_tests key'))
+
+
+def connect_ssh(vm, creds,
+                num_sec=CONNECT_RETRIES_TIMEOUT,
+                connect_timeout=CONNECT_TIMEOUT,
+                delay=CONNECT_SSH_DELAY):
+    """
+    This function attempts to connect all the IPs of the vm in several
+    rounds. After each round it will delay.
+    Once connective IP is found, it will return SSHClient with established
+    connection.
+    """
+
+    msg = "The num_sec is smaller then connect_timeout. This looks like a bug."
+    assert num_sec > connect_timeout, msg
+
+    timestamp = time.monotonic()
+    # Note we do need to be sure we get the fresh list of IPs for every cycle.
+    # The provider may be updating it.
+    for ip in _trying_fresh_ips(vm, delay=delay, rounds=-1):
```

Comment:
Why isn't this using `pick_responding_ip`, as that wraps this method anyway?

I am not a fan of the pattern here, setting up an infinite generator and then comparing timestamps. It serves almost the same exact purpose as a wait_for call.

A net check is also less costly than trying to establish the ssh connection immediately.

Reply:
### Wait for
I am not sure `wait_for` could be used. When I started this effort, I didn't have the `use_check_port` attribute. Perhaps now when it is implemented wait_for could be used. I will think this trough.

### Net check
How is the `net_check` less costly? What kind of resources you are talking about? I think the opposite is true. Checking with `net_check` means you need to be creating the connection twice -- once for checking, the other time for the connection that will be used for payload. So why to not just trying to connect with some timeout and try again if it timeouts? Is it because of the SSHClient's object creation costs?

If server is connective first time, it doesn't need to be the case the second time (when the payload-serving connection attempt is made) as for example some lousy firewall might start meanwhile or some transition effect (perhaps services deps cycles could be causing this) can happen. Once we have connection established, it may survive such things (for example ssh will survive restart of the sshd as it is not killing it's children). And yes, issues like that are real. I think services deps issues may be the case why I am getting connection_reset in some of my tests. One such test is included in this PR and makes PRT failures sometimes. It may be great to have a wrapper around the SSHClient to make it reliably reconnecting. It is not working fine now even though I think there was some effort to make it reliable.

Reply:
I am making this change now, and testing. I just did a simple replacement of the function we call and added the args. However, I believe this undoes some of the ability to pass values down that was initially intended. This was written by Jaroslav so I need to take a closer look and possibly get him involved.

Reply:
Testing is failing so I am debugging.

Reply:
So how do we want to proceed on this? I really thought we had come up with an appropriate solution. Seems like we need to nail down what are the specific needs/requirement that we are trying to accomplish with these updates to connect_ssh.